### PR TITLE
Updated to allow non-base64 values for EK and null/empty/non-base64 values for SRK.

### LINKED ISSUE #378: query.NextAsync() errors when the storage root key is empty string
- **OS and version used:** Windows 10 <VERSION> 

- **SDK version used:** latest solution from github and also the nuget package <VERSION> 

# Description of the issue:

I am using QuerySpecification for CreateIndividualEnrollmentQuery. When the enrollments are received from the api I get a error for TPM based devices. On debugging the code it seems that the storage root key is returned as empty string and when the json is deserialized using JsonConvert.DeserializeObject I get the error. The error occurs in TpmAttestation.cs file. Can anyone please help?

# Code sample exhibiting the issue:

Code snippet:
Console.WriteLine("\nCreating a query for enrollments...");
                QuerySpecification querySpecification = new QuerySpecification("SELECT * FROM enrollments where registrationId='dpspoc1'");
                using (Query query = provisioningServiceClient.CreateIndividualEnrollmentQuery(querySpecification, QueryPageSize))
                {
                    while (query.HasNext())
                    {
                        Console.WriteLine("\nQuerying the next enrollments...");
                        QueryResult queryResult = await query.NextAsync().ConfigureAwait(false);
                        Console.WriteLine(queryResult);
                    }
                }

Stack trace of error:

Microsoft.Azure.Devices.Provisioning.Service.ProvisioningServiceClientException
  HResult=0x80131500
  Message=
  Source=Microsoft.Azure.Devices.Provisioning.Service
  StackTrace:
   at Microsoft.Azure.Devices.Provisioning.Service.TpmAttestation..ctor(String endorsementKey, String storageRootKey) in C:\Amogh\POC\azure-iot-sdk-csharp\provisioning\service\src\Config\TpmAttestation.cs:line 45
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolvePropertyAndCreatorValues(JsonObjectContract contract, JsonProperty containerProperty, JsonReader reader, Type objectType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolvePropertyAndCreatorValues(JsonObjectContract contract, JsonProperty containerProperty, JsonReader reader, Type objectType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateList(IList list, JsonReader reader, JsonArrayContract contract, JsonProperty containerProperty, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateList(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, Object existingValue, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Microsoft.Azure.Devices.Provisioning.Service.QueryResult..ctor(String typeString, String bodyString, String continuationToken) in C:\Amogh\POC\azure-iot-sdk-csharp\provisioning\service\src\Config\QueryResult.cs:line 120
   at Microsoft.Azure.Devices.Provisioning.Service.Query.<NextAsync>d__20.MoveNext() in C:\Amogh\POC\azure-iot-sdk-csharp\provisioning\service\src\Query.cs:line 232
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at ProvisioningServiceBulkOperation.Program.<RunSample>d__8.MoveNext() in C:\Amogh\POC\azure-iot-sdk-csharp\provisioning\service\samples\ProvisioningServiceBulkOperation\Program.cs:line 99
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ProvisioningServiceBulkOperation.Program.Main(String[] args) in C:\Amogh\POC\azure-iot-sdk-csharp\provisioning\service\samples\ProvisioningServiceBulkOperation\Program.cs:line 52

Inner Exception 1:
ArgumentException: parameter cannot be null or whiteSpace

# Console log of the issue:

### PR DESCRIPTION
**Pull request by abhipsaMisra**
# Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [x] This pull-request is submitted against the `master` branch.

# Description of the changes

- Modified the DPS Service code to allow non-base64 values for Endorsement Key and null/empty/non-base64 values for Storage Root Key.

# Reference/Link to the issue solved with this PR (if any)
Fixes #378

### CODE REVIEW

File: provisioning/service/src/Config/ParserUtility.cs [hunk 1]
```diff
@@ -72,6 +72,23 @@ public static void EnsureBase64String(string input)
...
+        /// </summary>
+        /// <param name="input">the <code>string</code> to be validated</param>
+        /// <exception cref="ArgumentException">if the provided <code>string</code> do not fits the criteria</exception>
+        public static void EnsureNotNullOrWhiteSpaceString(string input)
+        {
+            if (string.IsNullOrWhiteSpace(input))
+            {
+                /* Codes_SRS_PARSER_UTILITY_21_016: [The IsNullOrWhiteSpace shall throw ArgumentException if the provided string is null.] */
+                /* Codes_SRS_PARSER_UTILITY_21_017: [The IsNullOrWhiteSpace shall throw ArgumentException if the provided string is whitespace.] */
+                throw new ArgumentException("parameter cannot be null or whiteSpace");
```
Comment (CIPop) [inline provisioning/service/src/Config/ParserUtility.cs]: Please move the check within the function requiring it and throw `ArgumentNullException` passing the `nameof(argument)`. 
`parameter` is too generic and is not in-line with C# argument validation.

File: provisioning/service/src/Config/TpmAttestation.cs [hunk 2]
```diff
@@ -59,8 +59,8 @@ public string EndorsementKey
             private set
             {
                 /* SRS_TPM_ATTESTATION_21_001: [The EndorsementKey setter shall throws ArgumentException if the provided 
-                                                endorsementKey is null or invalid.] */
-                ParserUtils.EnsureBase64String(value);
+                                                endorsementKey is null or white space.] */
+                ParserUtils.EnsureNotNullOrWhiteSpaceString(value);
```
Comment (CIPop) [inline provisioning/service/src/Config/TpmAttestation.cs]: `string` has this check. This is overly complicated and instead removes code clarity.

File: provisioning/service/tests/Config/ParserUtilsTests.cs [hunk 3]
```diff
@@ -132,5 +132,25 @@ public void ParserUtils_IsValidId_ThrowsOnInvalidId()
                 "12345678901234567890123456789012345678901234567890" +
                 "12345678901234567890123456789"));
         }
+
+        /* Codes_SRS_PARSER_UTILITY_21_015: [The IsNullOrWhiteSpace shall do nothing if the string is neither null nor white space.] */
+        [TestMethod]
+        [TestCategory("DevService")]
+        public void ParserUtils_IsNotNullOrWhiteSpace_SucceedOnNotNullOrWhiteSpace()
```
Comment (CIPop) [inline provisioning/service/tests/Config/ParserUtilsTests.cs]: Not needed - testing string.IsNullOrEmpty.

File: provisioning/service/tests/Config/TpmAttestationTests.cs [hunk 4]
```diff
@@ -30,7 +31,7 @@ public void TpmAttestation_ThrowsOnNullEndorsementKey()
 
         [TestMethod]
         [TestCategory("DevService")]
-        public void TpmAttestation_ThrowsOnEnptyEndorsementKey()
+        public void TpmAttestation_ThrowsOnEmptyEndorsementKey()
```
Comment (CIPop) [inline provisioning/service/tests/Config/TpmAttestationTests.cs]: Missing test for null.

File: provisioning/service/tests/Config/TpmAttestationTests.cs [hunk 5]
```diff
@@ -41,16 +42,20 @@ public void TpmAttestation_ThrowsOnEnptyEndorsementKey()
 
         [TestMethod]
         [TestCategory("DevService")]
-        public void TpmAttestation_ThrowsOnNonBase64EndorsementKey()
+        public void TpmAttestation_SucceedOnNonBase64EndorsementKey()
```
Comment (CIPop) [inline provisioning/service/tests/Config/TpmAttestationTests.cs]: Tests describe the requirements, they are not testing things that are _not_ required.

File: provisioning/service/tests/Config/TpmAttestationTests.cs [hunk 6]
```diff
@@ -69,26 +74,35 @@ public void TpmAttestation_SucceedOnNullStorageRootKey()
 
         [TestMethod]
         [TestCategory("DevService")]
-        public void TpmAttestation_ThrowsOnEnptyStorageRootKey()
+        public void TpmAttestation_SucceedOnEmptyStorageRootKey()
```
Comment (CIPop) [inline provisioning/service/tests/Config/TpmAttestationTests.cs]: Missing null test.

File: provisioning/service/tests/Config/TpmAttestationTests.cs [hunk 7]
```diff
@@ -69,26 +74,35 @@ public void TpmAttestation_SucceedOnNullStorageRootKey()
...
+            // assert
+            Assert.AreEqual(endorsementKey, tpmAttestation.EndorsementKey);
+            Assert.AreEqual(storageRootKey, tpmAttestation.StorageRootKey);
         }
 
+
         [TestMethod]
         [TestCategory("DevService")]
-        public void TpmAttestation_ThrowsOnNonBase64StorageRootKey()
+        public void TpmAttestation_SucceedOnNonBase64StorageRootKey()
```
Comment (CIPop) [inline provisioning/service/tests/Config/TpmAttestationTests.cs]: Same as above: please remove.

File: provisioning/service/tests/Config/TpmAttestationTests.cs [hunk 8]
```diff
@@ -215,5 +250,44 @@ public void TpmAttestation_ThrowsOnDeserializationWithEmptyEndorsementKey()
...
+
+            // assert
+            Assert.AreEqual(endorsementKey, tpmAttestation.EndorsementKey);
+        }
+
+        [TestMethod]
+        [TestCategory("DevService")]
+        public void TpmAttestation_SucceedOnDeserializationWithEmptyStorageRootKey()
+        {
+            // arrange
```
Comment (CIPop) [inline provisioning/service/tests/Config/TpmAttestationTests.cs]: Too much code duplication. Compact within a single test please.

File: provisioning/service/tests/Config/TpmAttestationTests.cs [hunk 9]
```diff
@@ -164,6 +178,27 @@ public void TpmAttestation_SucceedOnSerializationWithoutStorageRootKey()
             TestAssert.AreEqualJson(expectedJson, json);
         }
 
+        [TestMethod]
+        [TestCategory("DevService")]
+        public void TpmAttestation_SucceedOnSerializationWithEmptyStorageRootKey()
```
Comment (CIPop) [inline provisioning/service/tests/Config/TpmAttestationTests.cs]: Do we already have a test with a null/empty EK?

File: provisioning/service/tests/Config/TpmAttestationTests.cs (hunk 4 above)
Comment (abhipsaMisra) [reply provisioning/service/tests/Config/TpmAttestationTests.cs]: `TpmAttestationTests` has a test for null EK `TpmAttestation_ThrowsOnNullEndorsementKey`

File: provisioning/service/tests/Config/TpmAttestationTests.cs (hunk 6 above)
Comment (abhipsaMisra) [reply provisioning/service/tests/Config/TpmAttestationTests.cs]: `TpmAttestationTests` has a test for `TpmAttestation_SucceedOnNullStorageRootKey`

File: provisioning/service/tests/Config/TpmAttestationTests.cs (hunk 8 above)
Comment (abhipsaMisra) [reply provisioning/service/tests/Config/TpmAttestationTests.cs]: Testing for multiple scenarios with deserialization. Will look into using parameterized test in the future

File: provisioning/service/tests/Config/TpmAttestationTests.cs (hunk 9 above)
Comment (abhipsaMisra) [reply provisioning/service/tests/Config/TpmAttestationTests.cs]: `TpmAttestationTests` already has tests for null/empty EK. Since serialization can occur only after the TpmAttestation object has been created, null/empty EK would fail TpmAttestation object creation itself, thereby failing any subsequent actions such as serialization etc.

File: provisioning/service/src/Config/TpmAttestation.cs [hunk 10]
```diff
@@ -78,12 +81,7 @@ public string StorageRootKey
             }
             private set
             {
-                /* SRS_TPM_ATTESTATION_21_002: [The StorageRootKey setter shall throws ArgumentException if the provided 
-                                                storageRootKey is not null but invalid.] */
-                if (value != null)
-                {
-                    ParserUtils.EnsureBase64String(value);
```
Comment (CIPop) [inline provisioning/service/src/Config/TpmAttestation.cs]: Is this used anywhere else?
Can we remove it and the associated UTs?
